### PR TITLE
feat(sessiond): Extended UT coverage for sessiond

### LINKED ISSUE #8713: Implement more unit test cases for extra UT coverage in sessiond
Unit Test Cases to be added in sessiond for below features:
1) PDU session establishment
2) PDU session Release
3) Service Request.

### PR DESCRIPTION
**Pull request by GANESH-WAVELABS**
feat(sessiond): Extended UT coverage for sessiond

## Summary
i. fixed the UT file name by renaming as listed below:
image
ii. modified few test cases and added EXPECT/ASSERTIONS for the UT file "test_set_session_manager_handler.cpp"  
iii. Corresponding github issue is #8713

## Test Plan
image

## Additional Information
Signed-off-by: GANESH-WAVELABS <ganesh.irrinki@wavelabs.ai>

### CODE REVIEW

**Comment by uri200** [comment]
please, execute `precommit_sm` so it applies format to your changes

**Comment by uri200** [comment]
Also remember to add `EXPECT_CALL` to mock calls to external systems (like grcp services)

Here you have an example
sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=orc8r_add_demap_endpoints_to_directoryd&file=lte%2Fgateway%2Fc%2Fsession_manager%2Ftest%2Ftest_local_enforcer.cpp&editor=JetBrains&version=v1.2.1&utm_product_name=CLion&utm_product_version=2021.1.2&start_row=2158&start_col=20&end_row=2158&end_col=20

File: lte/gateway/c/session_manager/test/Consts.h [hunk 1]
```diff
@@ -43,3 +43,8 @@
 #define BEARER_ID_1 0
 #define BEARER_ID_2 2
 #define BEARER_ID_3 3
+#define UE_IPV4 "192.168.128.11"
+#define AGW_IPV4 "192.168.60.142"
+#define REDIRECT_SERVER_IPV4_ADDRESS "10.20.30.40"
+#define PDU_ID 0x5
+#define APN "facebook.com"
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/test/Consts.h]: Let's move this up to line 30 and name it APN2 or APN_FACEBOOK. Do we need these new IPV4 consts?

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 2]
```diff
@@ -100,24 +100,15 @@ class SessionManagerHandlerTest : public ::testing::Test {
...
+    req->mutable_gnode_endpoint()->set_teid(TEID_1_UL);
+    req->mutable_gnode_endpoint()->set_end_ipv4_addr(UE_IPV4);
+    req->mutable_upf_endpoint()->set_teid(TEID_1_DL);
+    req->mutable_upf_endpoint()->set_end_ipv4_addr(AGW_IPV4);
 
     reqcmn->mutable_sid()->set_id("IMSI000000000000001");
+    reqcmn->mutable_teids()->set_enb_teid(TEID_2_UL);
+    reqcmn->mutable_teids()->set_agw_teid(TEID_2_DL);
     reqcmn->set_sm_session_state(magma::SMSessionFSMState::CREATING_0);
-
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:106]: why are these removed?

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 3]
```diff
@@ -132,40 +123,38 @@ class SessionManagerHandlerTest : public ::testing::Test {
         magma::NotifyUeEvents::PDU_SESSION_INACTIVE_NOTIFY);
 
     reqcmn->mutable_sid()->set_id("IMSI000000000000001");
-    reqcmn->set_apn("BLR");
+    reqcmn->set_apn("facebook.com");
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp]: can you use  something different like `apn` or `apn.com` or `example.com`?

File: lte/gateway/c/session_manager/test/Consts.h (hunk 1 above)
Comment (uri200) [reply lte/gateway/c/session_manager/test/Consts.h]: and lets not use facebook.com

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 4]
```diff
@@ -132,40 +123,38 @@ class SessionManagerHandlerTest : public ::testing::Test {
...
+    EXPECT_EQ(request->mutable_common_context()->teids().agw_teid(), TEID_2_DL);
     EXPECT_EQ(
-        noti.notify_ue_event(), magma::NotifyUeEvents::UE_IDLE_MODE_NOTIFY);
-
-    set_session_manager->initiate_release_session(session_map_, pdu_id, imsi);
-    reqcmn->set_sm_session_state(magma::SMSessionFSMState::RELEASED_4);
-    EXPECT_EQ(reqcmn->sm_session_state(), magma::SMSessionFSMState::RELEASED_4);
-    EXPECT_EQ(reqcmn->sm_session_version(), 0);
+        request->mutable_common_context()->sm_session_state(),
+        magma::SMSessionFSMState::CREATING_0);
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:161]: why is imsi or apn never checked

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 5]
```diff
@@ -192,12 +181,24 @@ TEST_F(SessionManagerHandlerTest, test_SetAmfSessionContext) {
       &server_context, &request,
       [this](grpc::Status status, SmContextVoid Void) {});
 
+  assert_sm_session_context(&request);
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:195]: why is this assert check here instead of being checked right after `set_sm_session_context`

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 6]
```diff
@@ -192,12 +181,24 @@ TEST_F(SessionManagerHandlerTest, test_SetAmfSessionContext) {
...
+  EXPECT_EQ(request.mutable_common_context()->sid().id(), IMSI1);
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:196]: why is this checked outside `assert_sm_session_context`

File: lte/gateway/c/session_manager/test/test_upf_msg_manage_handler.cpp [hunk 7]
```diff
@@ -73,36 +96,130 @@ class SetUPFNodeState : public ::testing::Test {
...
+      break;
+    }
+    case UPFNodeState::kNodeReport: {
+      break;
+    }
+    case UPFNodeState::UPF_NODE_MESSAGES_NOT_SET: {
+      break;
+    }
+    default: { break; }
+  }
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_upf_msg_manage_handler.cpp:199]: I don;t understand well this switch. Normally a test is a lineal sequence of events and this seems to allow different cases. Can you explain what is this switch doing?

File: lte/gateway/c/session_manager/test/Consts.h (hunk 1 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/Consts.h]: Yes, renamed as APN2 "example.com"

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp (hunk 3 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp]: Yes, renamed as APN2 "example.com"

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp (hunk 2 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:106]: earlier `set_sm_session_context` function was having both set, assert functionalities. I have differed set and assert as two different functions.

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp (hunk 4 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:161]: Added assertion for apn in `assert_sm_session_context`. for imsi we were checking outside `assert_sm_session_context` as in some places we were using IMSI1, some places we were using IMSI2. so to pass those test cases I moved out imis checking out of `assert_sm_session_context` function.

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp (hunk 6 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:196]: for imsi we were checking outside `assert_sm_session_context` as in some places we were using IMSI1, some places we were using IMSI2. so to pass those test cases I moved out imis checking out of `assert_sm_session_context` function.

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp (hunk 5 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:195]: Though we are setting GRPC message `SetSMSessionContext` request locally via `set_sm_session_context` , we can assume the flow like, once AMF fills the `SetSMSessionContext` request and sends to sessiond via GRPC call `SetAmfSessionContext` and then we can assert via `assert_sm_session_context` function. I thought it would be more readable. Please suggest me if it is fine or not.

**Comment by GANESH-WAVELABS** [comment]
> Also remember to add `EXPECT_CALL` to mock calls to external systems (like grcp services)
> 
> Here you have an example sourcegraph.com/-/editor?remote_url=git%40github.com%3Auri200%2Fmagma.git&branch=orc8r_add_demap_endpoints_to_directoryd&file=lte%2Fgateway%2Fc%2Fsession_manager%2Ftest%2Ftest_local_enforcer.cpp&editor=JetBrains&version=v1.2.1&utm_product_name=CLion&utm_product_version=2021.1.2&start_row=2158&start_col=20&end_row=2158&end_col=20

`UpfMsgManageHandler.h` file is a GRPC receiver for pipelined responses towards sessiond
`SetMessageManagerHandler.h` file is a GRPC receiver for MME/AMF settings towards sessiond
so we can not create EXPECT_CALL() on GRPC receiver class. So could not add EXPECT_CALL() in these classes test cases.

File: lte/gateway/c/session_manager/test/test_upf_msg_manage_handler.cpp [hunk 8]
```diff
@@ -128,7 +245,20 @@ TEST_F(SetUPFNodeState, test_upf_session_config) {
      * with SMF latest version
      */
     EXPECT_LE(version, cur_version);
+
+    EXPECT_TRUE(
+        session_enforcer->is_incremented_rtx_counter_within_max(session));
+
     RulesToProcess pending_activation, pending_deactivation;
+    const CreateSessionResponse& csr = session->get_create_session_response();
+    std::vector<StaticRuleInstall> static_rule_installs =
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/test/test_upf_msg_manage_handler.cpp:254]: Can you also rename the file here as well? https://github.com/magma/magma/blob/master/lte/gateway/c/session_manager/test/BUILD.bazel#L261

You can verify the Bazel testing by running this in the Magm VM.

```bash
cd magma/
bazel test //lte/gateway/c/session_manager/test:all
```

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 9]
```diff
@@ -132,40 +127,45 @@ class SessionManagerHandlerTest : public ::testing::Test {
...
-    auto pdu_id = req->pdu_session_id();
-    auto cfg = set_session_manager->m5g_build_session_config(*session_ctx_req);
-    auto& noti    = *req;
-    auto ue_event = noti.notify_ue_event();
-    std::function<void(Status, SmContextVoid)> response_callback;
-
-    EXPECT_EQ(pdu_id, 0x5);
-    EXPECT_EQ(imsi, "IMSI000000000000001");
-    EXPECT_EQ(apn, "BLR");
+  void assert_sm_session_context(magma::SetSMSessionContext* request) {
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:136]: I am still confused with what are we really asserting here. Why do we need a test to assert an object we created on a test isself? 

In my opinion this test case it is only testing Googles GRPC code, but not magma.

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp (hunk 5 above)
Comment (uri200) [reply lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:195]: as mentioned before, assert_sm_session_context is testing an object created by this same test. It looks a test of a test. In my opinion this does not add any kind of coverage to our code

File: lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp [hunk 10]
```diff
@@ -192,12 +192,24 @@ TEST_F(SessionManagerHandlerTest, test_SetAmfSessionContext) {
...
+  cfg.common_context       = request.common_context();
+  cfg.rat_specific_context = request.rat_specific_context();
+  cfg.rat_specific_context.mutable_m5gsm_session_context()->set_ssc_mode(
+      SSC_MODE_3);
 
   auto session_map = session_store->read_sessions({IMSI1});
   auto it          = session_map.find(IMSI1);
   EXPECT_FALSE(it == session_map.end());
+  auto pdu_id = req->pdu_session_id();
+  set_session_manager->send_create_session(session_map, IMSI1, cfg, pdu_id);
```
Comment (uri200) [inline lte/gateway/c/session_manager/test/test_set_session_manager_handler.cpp:212]: In my opinion, this testcase can do much more: why aren't we checking every single parameter of the session just created.

Note to be effective, unit test must check the result values to see if what you created is what you expect. 

This applies for all the tests in this PR or session manager.

Note this is the contrary of `assert_sm_session_context`. In this case we do need to assert the content of the session to make sure it is equal to what we requested to be

File: lte/gateway/c/session_manager/test/test_upf_msg_manage_handler.cpp (hunk 8 above)
Comment (GANESH-WAVELABS) [reply lte/gateway/c/session_manager/test/test_upf_msg_manage_handler.cpp:254]: renamed the file `test_upf_node_state.cpp` to `test_upf_msg_manage_handler.cpp` under `magma/lte/gateway/c/session_manager/test/BUILD.bazel` . Also added one more UT file `test_session_state_5g.cpp` which got missed under `magma/lte/gateway/c/session_manager/test/BUILD.bazel`. Below is the screen shot for passing test cases for Bazel.
image

**Comment by uri200** [comment]
Hey @GANESH-WAVELABS  will you retake this pr?
